### ISSUE TITLE
Shlink `3.5.0` not working after update if using redis-sentinel

### ISSUE THREAD

**Issue by onedr0p** [shlinkio/shlink#1684]
#### Temporary workaround (from maintainer)

Until a proper solution has been put in place, if you are using redis sentinels and experiencing this error after updating to Shlink 3.5.0, you can work around it by deleting all entries in your redis cluster which key starts with `Shlink:`.

---

#### How Shlink is set-up

* Shlink Version: 3.5.0
* How do you serve Shlink: Docker image
* Database engine used: PostgreSQL

Configuration:

https://github.com/onedr0p/home-ops/tree/main/kubernetes/apps/default/shlink/api

#### Summary

Visiting a short link gives me this error `An unexpected error occurred`

```
✖ k logs shlink-api-86f77d97d8-bglk7 -f
Defaulted container "shlink-api" out of: shlink-api, init-db (init)
Creating fresh database if needed...
Updating database...
Generating proxies...
Clearing entities cache...
Downloading GeoLite2 db file...
Configuring periodic visit location...
[2023-01-28T15:23:42.464013-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 16
[2023-01-28T15:23:42.465896-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 21
[2023-01-28T15:23:42.471205-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 24
[2023-01-28T15:23:42.478998-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 18
[2023-01-28T15:23:42.481440-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 20
[2023-01-28T15:23:42.487144-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 23
[2023-01-28T15:23:42.471075-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 19
[2023-01-28T15:23:42.487995-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 22
[2023-01-28T15:23:42.489831-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 29
[2023-01-28T15:23:42.491763-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 0
[2023-01-28T15:23:42.493623-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 1
[2023-01-28T15:23:42.494329-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 27
[2023-01-28T15:23:42.497235-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 17
[2023-01-28T15:23:42.499051-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 26
[2023-01-28T15:23:42.492935-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 25
[2023-01-28T15:23:42.495756-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 28
[2023-01-28T15:23:42.501179-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 30
[2023-01-28T15:23:42.503097-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 2
[2023-01-28T15:23:42.499678-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 31
[2023-01-28T15:23:42.504903-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 3
[2023-01-28T15:23:42.507018-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 5
[2023-01-28T15:23:42.508858-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 6
[2023-01-28T15:23:42.509218-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 4
[2023-01-28T15:23:42.511054-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 12
[2023-01-28T15:23:42.513190-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 10
[2023-01-28T15:23:42.513188-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 9
[2023-01-28T15:23:42.514952-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 11
[2023-01-28T15:23:42.516087-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 15
[2023-01-28T15:23:42.516759-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 13
[2023-01-28T15:23:42.517100-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 14
[2023-01-28T15:23:42.518553-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 8
[2023-01-28T15:23:42.520234-05:00] [NULL] Access.NOTICE - Worker started in /etc/shlink with ID 7
[2023-01-28T15:24:09.868551-05:00] [NULL] Shlink.NOTICE - Starting work on task 0 using: {"handler":"Shlinkio\\Shlink\\Core\\EventDispatcher\\Mercure\\NotifyVisitToMercure","arguments":[{"visitId":"1649"}]}
[2023-01-28T15:24:09.873393-05:00] [NULL] Shlink.NOTICE - Starting work on task 1 using: {"handler":"Shlinkio\\Shlink\\Core\\EventDispatcher\\RabbitMq\\NotifyVisitToRabbitMq","arguments":[{"visitId":"1649"}]}
[2023-01-28T15:24:09.873664-05:00] [NULL] Shlink.NOTICE - Starting work on task 2 using: {"handler":"Shlinkio\\Shlink\\Core\\EventDispatcher\\RedisPubSub\\NotifyVisitToRedis","arguments":[{"visitId":"1649"}]}
[2023-01-28T15:24:09.881031-05:00] [NULL] Shlink.NOTICE - Starting work on task 3 using: {"handler":"Shlinkio\\Shlink\\Core\\EventDispatcher\\NotifyVisitToWebHooks","arguments":[{"visitId":"1649"}]}
[2023-01-28T15:24:09.888499-05:00] [NULL] Shlink.NOTICE - Starting work on task 4 using: {"handler":"Shlinkio\\Shlink\\Core\\EventDispatcher\\UpdateGeoLiteDb","arguments":[{"visitId":"1649"}]}
[2023-01-28T15:24:09.899216-05:00] [9e3cb1bdf89604e072267152b902c038] Shlink.ERROR - Error: Typed property Shlinkio\Shlink\Core\ShortUrl\Entity\ShortUrl::$deviceLongUrls must not be accessed before initialization in /etc/shlink/module/Core/src/ShortUrl/Entity/ShortUrl.php:203
Stack trace:
#0 /etc/shlink/module/Core/src/ShortUrl/Helper/ShortUrlRedirectionBuilder.php(30): Shlinkio\Shlink\Core\ShortUrl\Entity\ShortUrl->longUrlForDevice(Object(Shlinkio\Shlink\Core\Model\DeviceType))
#1 /etc/shlink/module/Core/src/Action/RedirectAction.php(29): Shlinkio\Shlink\Core\ShortUrl\Helper\ShortUrlRedirectionBuilder->buildShortUrlRedirect(Object(Shlinkio\Shlink\Core\ShortUrl\Entity\ShortUrl), Object(Laminas\Diactoros\ServerRequest))
#2 /etc/shlink/module/Core/src/Action/AbstractTrackingAction.php(35): Shlinkio\Shlink\Core\Action\RedirectAction->createSuccessResp(Object(Shlinkio\Shlink\Core\ShortUrl\Entity\ShortUrl), Object(Laminas\Diactoros\ServerRequest))
#3 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(27): Shlinkio\Shlink\Core\Action\AbstractTrackingAction->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#4 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#5 /etc/shlink/module/Core/src/ShortUrl/Middleware/TrimTrailingSlashMiddleware.php(25): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#6 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(27): Shlinkio\Shlink\Core\ShortUrl\Middleware\TrimTrailingSlashMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#7 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#8 /etc/shlink/vendor/akrabat/ip-address-middleware/src/IpAddress.php(132): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#9 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(27): RKA\Middleware\IpAddress->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#10 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#11 /etc/shlink/vendor/laminas/laminas-stratigility/src/MiddlewarePipe.php(75): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#12 /etc/shlink/vendor/mezzio/mezzio-router/src/Route.php(79): Laminas\Stratigility\MiddlewarePipe->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#13 /etc/shlink/vendor/mezzio/mezzio-router/src/RouteResult.php(109): Mezzio\Router\Route->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#14 /etc/shlink/vendor/mezzio/mezzio-router/src/Middleware/DispatchMiddleware.php(32): Mezzio\Router\RouteResult->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#15 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(27): Mezzio\Router\Middleware\DispatchMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#16 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#17 /etc/shlink/vendor/laminas/laminas-stratigility/src/MiddlewarePipe.php(75): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#18 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Laminas\Stratigility\MiddlewarePipe->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#19 /etc/shlink/vendor/laminas/laminas-stratigility/src/Middleware/PathMiddlewareDecorator.php(42): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#20 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Laminas\Stratigility\Middleware\PathMiddlewareDecorator->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#21 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(42): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#22 /etc/shlink/vendor/mezzio/mezzio-router/src/Middleware/ImplicitHeadMiddleware.php(72): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#23 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(27): Mezzio\Router\Middleware\ImplicitHeadMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#24 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#25 /etc/shlink/vendor/mezzio/mezzio-router/src/Middleware/RouteMiddleware.php(50): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#26 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(27): Mezzio\Router\Middleware\RouteMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#27 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#28 /etc/shlink/vendor/laminas/laminas-stratigility/src/MiddlewarePipe.php(75): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#29 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Laminas\Stratigility\MiddlewarePipe->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#30 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(42): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#31 /etc/shlink/vendor/shlinkio/shlink-common/src/Middleware/CloseDbConnectionMiddleware.php(24): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#32 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(27): Shlinkio\Shlink\Common\Middleware\CloseDbConnectionMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#33 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#34 /etc/shlink/vendor/laminas/laminas-stratigility/src/MiddlewarePipe.php(75): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#35 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Laminas\Stratigility\MiddlewarePipe->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#36 /etc/shlink/vendor/laminas/laminas-stratigility/src/Middleware/PathMiddlewareDecorator.php(42): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#37 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Laminas\Stratigility\Middleware\PathMiddlewareDecorator->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#38 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(42): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#39 /etc/shlink/module/Rest/src/Middleware/CrossDomainMiddleware.php(24): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#40 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(27): Shlinkio\Shlink\Rest\Middleware\CrossDomainMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#41 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#42 /etc/shlink/vendor/laminas/laminas-stratigility/src/Middleware/ErrorHandler.php(129): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#43 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(27): Laminas\Stratigility\Middleware\ErrorHandler->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#44 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#45 /etc/shlink/vendor/php-middleware/request-id/src/RequestIdMiddleware.php(32): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#46 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(27): PhpMiddleware\RequestId\RequestIdMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#47 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#48 /etc/shlink/vendor/shlinkio/shlink-common/src/Middleware/ContentLengthMiddleware.php(24): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#49 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(27): Shlinkio\Shlink\Common\Middleware\ContentLengthMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#50 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#51 /etc/shlink/vendor/laminas/laminas-stratigility/src/MiddlewarePipe.php(75): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#52 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(49): Laminas\Stratigility\MiddlewarePipe->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
#53 /etc/shlink/vendor/laminas/laminas-stratigility/src/MiddlewarePipe.php(75): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
#54 /etc/shlink/vendor/laminas/laminas-stratigility/src/MiddlewarePipe.php(64): Laminas\Stratigility\MiddlewarePipe->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\EmptyPipelineHandler))
#55 /etc/shlink/vendor/mezzio/mezzio-swoole/src/Event/RequestHandlerRequestListener.php(109): Laminas\Stratigility\MiddlewarePipe->handle(Object(Laminas\Diactoros\ServerRequest))
#56 /etc/shlink/vendor/mezzio/mezzio-swoole/src/Event/EventDispatcher.php(36): Mezzio\Swoole\Event\RequestHandlerRequestListener->__invoke(Object(Mezzio\Swoole\Event\RequestEvent))
#57 /etc/shlink/vendor/mezzio/mezzio-swoole/src/SwooleRequestHandlerRunner.php(147): Mezzio\Swoole\Event\EventDispatcher->dispatch(Object(Mezzio\Swoole\Event\RequestEvent))
#58 [internal function]: Mezzio\Swoole\SwooleRequestHandlerRunner->onRequest(Object(Swoole\Http\Request), Object(Swoole\Http\Response))
#59 {main}
[2023-01-28T15:24:09.900543-05:00] [9e3cb1bdf89604e072267152b902c038] Access.ERROR - - "GET /tu0B6 HTTP/1.1" 500 28
[2023-01-28T15:24:09.946953-05:00] [9e3cb1bdf89604e072267152b902c038] Shlink.NOTICE - Task 3 has finished
[2023-01-28T15:24:09.963695-05:00] [9e3cb1bdf89604e072267152b902c038] Shlink.NOTICE - Task 4 has finished
[2023-01-28T15:24:09.965457-05:00] [9e3cb1bdf89604e072267152b902c038] Shlink.NOTICE - Task 1 has finished
[2023-01-28T15:24:09.983032-05:00] [NULL] Shlink.ERROR - Error processing task 0: [Error - 0] Typed property Shlinkio\Shlink\Core\ShortUrl\Entity\ShortUrl::$deviceLongUrls must not be accessed before initialization
#0 /etc/shlink/module/Core/src/ShortUrl/Transformer/ShortUrlDataTransformer.php(30): Shlinkio\Shlink\Core\ShortUrl\Entity\ShortUrl->deviceLongUrls()
#1 /etc/shlink/module/Core/src/EventDispatcher/PublishingUpdatesGenerator.php(41): Shlinkio\Shlink\Core\ShortUrl\Transformer\ShortUrlDataTransformer->transform(Object(DoctrineProxies\__CG__\Shlinkio\Shlink\Core\ShortUrl\Entity\ShortUrl))
#2 /etc/shlink/module/Core/src/EventDispatcher/Async/AbstractNotifyVisitListener.php(68): Shlinkio\Shlink\Core\EventDispatcher\PublishingUpdatesGenerator->newShortUrlVisitUpdate(Object(Shlinkio\Shlink\Core\Visit\Entity\Visit))
#3 /etc/shlink/module/Core/src/EventDispatcher/Async/AbstractNotifyVisitListener.php(46): Shlinkio\Shlink\Core\EventDispatcher\Async\AbstractNotifyVisitListener->determineUpdatesForVisit(Object(Shlinkio\Shlink\Core\Visit\Entity\Visit))
#4 /etc/shlink/module/Core/src/EventDispatcher/CloseDbConnectionEventListener.php(24): Shlinkio\Shlink\Core\EventDispatcher\Async\AbstractNotifyVisitListener->__invoke(Object(Shlinkio\Shlink\Core\EventDispatcher\Event\VisitLocated))
#5 /etc/shlink/vendor/mezzio/mezzio-swoole/src/Task/ServiceBasedTask.php(52): Shlinkio\Shlink\Core\EventDispatcher\CloseDbConnectionEventListener->__invoke(Object(Shlinkio\Shlink\Core\EventDispatcher\Event\VisitLocated))
#6 /etc/shlink/vendor/mezzio/mezzio-swoole/src/Task/TaskInvokerListener.php(45): Mezzio\Swoole\Task\ServiceBasedTask->__invoke(Object(Laminas\ServiceManager\ServiceManager))
#7 /etc/shlink/vendor/mezzio/mezzio-swoole/src/Event/EventDispatcher.php(36): Mezzio\Swoole\Task\TaskInvokerListener->__invoke(Object(Mezzio\Swoole\Event\TaskEvent))
#8 /etc/shlink/vendor/mezzio/mezzio-swoole/src/SwooleRequestHandlerRunner.php(197): Mezzio\Swoole\Event\EventDispatcher->dispatch(Object(Mezzio\Swoole\Event\TaskEvent))
#9 [internal function]: Mezzio\Swoole\SwooleRequestHandlerRunner->onTask(Object(Swoole\Http\Server), 0, 8, Object(Mezzio\Swoole\Task\ServiceBasedTask))
#10 /etc/shlink/vendor/mezzio/mezzio-swoole/src/SwooleRequestHandlerRunner.php(91): Swoole\Server->start()
#11 /etc/shlink/vendor/mezzio/mezzio/src/Application.php(53): Mezzio\Swoole\SwooleRequestHandlerRunner->run()
#12 /etc/shlink/vendor/mezzio/mezzio-swoole/src/Command/StartCommand.php(138): Mezzio\Application->run()
#13 /etc/shlink/vendor/symfony/console/Command/Command.php(312): Mezzio\Swoole\Command\StartCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#14 /etc/shlink/vendor/symfony/console/Application.php(1040): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#15 /etc/shlink/vendor/symfony/console/Application.php(314): Symfony\Component\Console\Application->doRunCommand(Object(Mezzio\Swoole\Command\StartCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#16 /etc/shlink/vendor/symfony/console/Application.php(168): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#17 /etc/shlink/vendor/laminas/laminas-cli/bin/laminas(54): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#18 /etc/shlink/vendor/bin/laminas(120): include('/etc/shlink/ven...')
#19 {main}
[2023-01-28T15:24:09.984052-05:00] [9e3cb1bdf89604e072267152b902c038] Shlink.NOTICE - Task 0 has finished
[2023-01-28T15:24:09.988696-05:00] [NULL] Shlink.ERROR - Error processing task 2: [Error - 0] Typed property Shlinkio\Shlink\Core\ShortUrl\Entity\ShortUrl::$deviceLongUrls must not be accessed before initialization
#0 /etc/shlink/module/Core/src/ShortUrl/Transformer/ShortUrlDataTransformer.php(30): Shlinkio\Shlink\Core\ShortUrl\Entity\ShortUrl->deviceLongUrls()
#1 /etc/shlink/module/Core/src/EventDispatcher/PublishingUpdatesGenerator.php(41): Shlinkio\Shlink\Core\ShortUrl\Transformer\ShortUrlDataTransformer->transform(Object(DoctrineProxies\__CG__\Shlinkio\Shlink\Core\ShortUrl\Entity\ShortUrl))
#2 /etc/shlink/module/Core/src/EventDispatcher/Async/AbstractNotifyVisitListener.php(68): Shlinkio\Shlink\Core\EventDispatcher\PublishingUpdatesGenerator->newShortUrlVisitUpdate(Object(Shlinkio\Shlink\Core\Visit\Entity\Visit))
#3 /etc/shlink/module/Core/src/EventDispatcher/Async/AbstractNotifyVisitListener.php(46): Shlinkio\Shlink\Core\EventDispatcher\Async\AbstractNotifyVisitListener->determineUpdatesForVisit(Object(Shlinkio\Shlink\Core\Visit\Entity\Visit))
#4 /etc/shlink/module/Core/src/EventDispatcher/CloseDbConnectionEventListener.php(24): Shlinkio\Shlink\Core\EventDispatcher\Async\AbstractNotifyVisitListener->__invoke(Object(Shlinkio\Shlink\Core\EventDispatcher\Event\VisitLocated))
#5 /etc/shlink/vendor/mezzio/mezzio-swoole/src/Task/ServiceBasedTask.php(52): Shlinkio\Shlink\Core\EventDispatcher\CloseDbConnectionEventListener->__invoke(Object(Shlinkio\Shlink\Core\EventDispatcher\Event\VisitLocated))
#6 /etc/shlink/vendor/mezzio/mezzio-swoole/src/Task/TaskInvokerListener.php(45): Mezzio\Swoole\Task\ServiceBasedTask->__invoke(Object(Laminas\ServiceManager\ServiceManager))
#7 /etc/shlink/vendor/mezzio/mezzio-swoole/src/Event/EventDispatcher.php(36): Mezzio\Swoole\Task\TaskInvokerListener->__invoke(Object(Mezzio\Swoole\Event\TaskEvent))
#8 /etc/shlink/vendor/mezzio/mezzio-swoole/src/SwooleRequestHandlerRunner.php(197): Mezzio\Swoole\Event\EventDispatcher->dispatch(Object(Mezzio\Swoole\Event\TaskEvent))
#9 [internal function]: Mezzio\Swoole\SwooleRequestHandlerRunner->onTask(Object(Swoole\Http\Server), 2, 8, Object(Mezzio\Swoole\Task\ServiceBasedTask))
#10 /etc/shlink/vendor/mezzio/mezzio-swoole/src/SwooleRequestHandlerRunner.php(91): Swoole\Server->start()
#11 /etc/shlink/vendor/mezzio/mezzio/src/Application.php(53): Mezzio\Swoole\SwooleRequestHandlerRunner->run()
#12 /etc/shlink/vendor/mezzio/mezzio-swoole/src/Command/StartCommand.php(138): Mezzio\Application->run()
#13 /etc/shlink/vendor/symfony/console/Command/Command.php(312): Mezzio\Swoole\Command\StartCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#14 /etc/shlink/vendor/symfony/console/Application.php(1040): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#15 /etc/shlink/vendor/symfony/console/Application.php(314): Symfony\Component\Console\Application->doRunCommand(Object(Mezzio\Swoole\Command\StartCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#16 /etc/shlink/vendor/symfony/console/Application.php(168): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#17 /etc/shlink/vendor/laminas/laminas-cli/bin/laminas(54): Symfony\Component\Console\Application->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#18 /etc/shlink/vendor/bin/laminas(120): include('/etc/shlink/ven...')
#19 {main}
[2023-01-28T15:24:09.988915-05:00] [9e3cb1bdf89604e072267152b902c038] Shlink.NOTICE - Task 2 has finished
```

**Comment by onedr0p** [comment]
I've also noticed the shlink-web is not able to load the short links.

**Comment by acelaya** [comment]
> I've also noticed the shlink-web is not able to load the short links.

I'll split these into two issues, as the second one is probably not related

**Comment by acelaya** [comment]
@onedr0p which docker image are you using? openswoole one or roadrunner one?

**Comment by acelaya** [comment]
> @onedr0p which docker image are you using? openswoole one or roadrunner one?

Ok, stupid question. Based on the shared logs it's clearly openswoole 😅

**Comment by acelaya** [comment]
@onedr0p I'm struggling to reproduce this. Does it happen consistently with every short URL? Only with new ones? Only with old ones?

Do you have any integrations enabled? The logs suggest you use mercure and redis pub/sub for real time updates. Is that correct?

Can you share all the env vars you pass to Shlink? (make sure to redact credentials)

**Comment by acelaya** [comment]
I have a theory. I think the problem is that there's outdated database info in redis, which is resulting in the ORM not knowing about the existence of the new `deviceLongUrls` prop, and therefore, it does not get initialized.

That is what is causing the `Typed property Shlinkio\Shlink\Core\ShortUrl\Entity\ShortUrl::$deviceLongUrls must not be accessed before initialization` errors, but also not making it reproducible without redis integration.

In theory, all the cache in redis related with the database gets cleared on container startup, precisely to avoid this, but something is not working there.

I'll do a couple more tests.

**Comment by onedr0p** [comment]
My env configuration

https://github.com/onedr0p/home-ops/blob/main/kubernetes/apps/default/shlink/api/helmrelease.yaml#L45-L54

https://github.com/onedr0p/home-ops/blob/main/kubernetes/apps/default/shlink/api/externalsecret.yaml#L19-L24

**Comment by acelaya** [comment]
I have tried to reproduce it doing this:

* Set-up an environment with Shlink 3.3.0 + Redis (Pub/Sub enabled) + Postgres
* Create a couple of short URLs
* Update to Shlink 3.4.0 (which introduced a change in cache management)
  * Everything still working
* Updated to Shlink 3.5.0
  * Verified the cache was properly cleared on container start-up
  * Everything was still working as expoected

Still I think the problem is that there's some outdated metadata cache somewhere.

I will try to set-up a redis cluster with sentinels, to make it 100% the same. Maybe the library abstracting the integration with redis is not properly clearing the entries on every instance.

In the meantime, could you try to manually delete all entries in redis which key starts with `Shlink:`? If that solves the problem, that would confirm my theory.

**Comment by onedr0p** [comment]
> In the meantime, could you try to manually delete all entries in redis which key starts with Shlink:? If that solves the problem, that would confirm my theory.

I figured this is where all this was heading 😃 . After I deleted all the `Shlink:*` keys, everything is happy again.

**Comment by acelaya** [comment]
Thanks for confirming! This can at least be a temporary workaround for those affected.

I'll try to reproduce the same scenario and see what I find.

**Comment by acelaya** [comment]
Ok, it has been hard, but I have been able to reproduce it.

As I suspected, it happens only when using redis-sentinel, where for some reason, the cache clearing ends in no entries being deleted. This leads to the error reported here.

I have run my Shlink with `SHELL_VERBOSITY=3`, and this is printed when the container starts-up:

```
Creating fresh database if needed...

 [OK] Database already exists. Run "db:migrate" command to make sure it is up to
      date.                                                                     

Updating database...
Migrating database...
  RUN  '/usr/local/bin/php' 'vendor/doctrine/migrations/bin/doctrine-migrations.php' 'migrations:migrate' '--no-interaction'
  OUT  
  OUT  
  RES  Command ran successfully

 [OK] Database properly migrated!                                               

Generating proxies...

 Processing entity "Shlinkio\Shlink\Core\Tag\Entity\Tag"
 Processing entity "Shlinkio\Shlink\Core\Visit\Entity\Visit"
 Processing entity "Shlinkio\Shlink\Core\ShortUrl\Entity\ShortUrl"
 Processing entity "Shlinkio\Shlink\Core\Visit\Entity\VisitLocation"
 Processing entity "Shlinkio\Shlink\Core\ShortUrl\Entity\DeviceLongUrl"
 Processing entity "Shlinkio\Shlink\Core\Domain\Entity\Domain"
 Processing entity "Shlinkio\Shlink\Rest\Entity\ApiKeyRole"
 Processing entity "Shlinkio\Shlink\Rest\Entity\ApiKey"

 Proxy classes generated to "/etc/shlink/data/proxies"
Clearing entities cache...

 // Clearing all Metadata cache entries                                         

 [OK] No cache entries were deleted.                                            
```

Notice the last line, `[OK] No cache entries were deleted.`. That should not be the case when updating from Shlink 3.4.0

I also don't know if this happens now because of [this change](https://github.com/shlinkio/shlink/pull/1623) introduced in Shlink 3.4.0, to make sure Shlink only clears its own entries in the redis cluster, or if it would have happened with the previous approach in which a full `FLUSHDB` was executed.

My suspicion is that there's a bug in some of the libraries Shlink uses, so I now need to track it down in order to repor it.

It's not easy. The metada cache clearing happens via a command of the database abstraction, which in turn uses a specific cache library, which in turn uses a specific library to communicate with redis. The bug can be in any of the three 😅

**Comment by acelaya** [comment]
Ok, I found the root cause.

Shlink uses predis, a library to connect to redis, and then another library which abstracts different types of cache (as redis is just one of the supported by Shlink).

The second one supports predis 1.*, but not long ago I updated Shlink to predis 2.*.

Everything seemed to work, except when you try to clear cache while using redis sentinels, as it calls a method that no longer exists. The second lib captures that error and just reports that cache was not cleared, but exits as successful.

I have opened an issue on that library, to request support for predis 2, and offering my help. However, that can take some time, so I have found a naive way to work around this and make it work for Shlink until the other library supports predis 2.

I have also considered downgrading to predis 1, but that's more work and the risk is higher.

**Comment by acelaya** [comment]
I have just released Shlink 3.5.1, which includes the fix. The docker images will be available once this is finished https://github.com/shlinkio/shlink/actions/runs/4092549701

**Comment by onedr0p** [comment]
Nice, glad you were able to get the bottom of this. thank you so much!
